kyototycoon/man/ktutilserv.1
.TH "KTUTILSERV" 1 "2012-05-25" "Man Page" "Kyoto Tycoon"

.SH NAME
ktutilserv \- server implementations to test miscellaneous utilities

.SH DESCRIPTION
.PP
The command `\fBktutilserv\fR' is a utility for facility test and performance test of the server tool kit.  This command is used in the following format.  `\fIbasedir\fR' specifies the document root directory.
.PP
.RS
.br
\fBktutilserv echo \fR[\fB\-host \fIstr\fB\fR]\fB \fR[\fB\-port \fInum\fB\fR]\fB \fR[\fB\-tout \fInum\fB\fR]\fB\fR
.RS
Runs the single\-threaded echo server.
.RE
.br
\fBktutilserv mtecho \fR[\fB\-host \fIstr\fB\fR]\fB \fR[\fB\-port \fInum\fB\fR]\fB \fR[\fB\-tout \fInum\fB\fR]\fB \fR[\fB\-th \fInum\fB\fR]\fB \fR[\fB\-li\fR|\fB\-ls\fR|\fB\-le\fR|\fB\-lz\fR]\fB\fR
.RS
Runs the multi\-threaded echo server.
.RE
.br
\fBktutilserv http \fR[\fB\-host \fIstr\fB\fR]\fB \fR[\fB\-port \fInum\fB\fR]\fB \fR[\fB\-tout \fInum\fB\fR]\fB \fR[\fB\-th \fInum\fB\fR]\fB \fR[\fB\-li\fR|\fB\-ls\fR|\fB\-le\fR|\fB\-lz\fR]\fB \fR[\fB\fIbasedir\fB\fR]\fB\fR
.RS
Runs the HTTP server.
.RE
.br
\fBktutilserv rpc \fR[\fB\-host \fIstr\fB\fR]\fB \fR[\fB\-port \fInum\fB\fR]\fB \fR[\fB\-tout \fInum\fB\fR]\fB \fR[\fB\-th \fInum\fB\fR]\fB \fR[\fB\-li\fR|\fB\-ls\fR|\fB\-le\fR|\fB\-lz\fR]\fB\fR
.RS
Runs the RPC server.
.RE
.RE
.PP
Options feature the following.
.PP
.RS
\fB\-host \fIstr\fR\fR : specifies the host name of the server.
.br
\fB\-port \fInum\fR\fR : specifies the port number of the server.
.br
\fB\-tout \fInum\fR\fR : specifies the timeout in seconds.
.br
\fB\-th \fInum\fR\fR : specifies the number of worker threads.  By default, it is 8.
.br
\fB\-li\fR : sets the logging level "INFO".
.br
\fB\-ls\fR : sets the logging level "SYSTEM".
.br
\fB\-le\fR : sets the logging level "ERROR".
.br
\fB\-lz\fR : sets the logging level "NONE".
.br
.RE
.PP
This command returns 0 on success, another on failure.

.SH SEE ALSO
.PP
.BR ktutiltest (1),
.BR ktutilmgr (1)


kyototycoon/ChangeLog
2012-05-01  FAL Labs  <info@fallabs.com>

	* cmdcommon.h (DBUpdateLogger::set_rsid): supressed a warning.

	* kttimeddb.h (TimedDB::check): new function.

	* ktremovedb.h (RemoteDB::check): new function.

	- Release: 0.9.56

2012-03-27  FAL Labs  <info@fallabs.com>

	* ktserver.c (Worker::do_synchronize): a source of the timestamp was added.

	- Release: 0.9.55

2012-03-16  FAL Labs  <info@fallabs.com>

	* ktserver.cc (Worker::do_match_similar): new function.

	* ktremotedb.h (RemoteDB::match_similar): new function.

	* kttimeddb.h (TimedDB::match_similar): new function.

	- Release: 0.9.54

2012-01-30  FAL Labs  <info@fallabs.com>

	* ktsocket.cc (setsocketoptions): omit error check because Solaris doesn't implement some options.

	- Release: 0.9.53

2011-10-30  FAL Labs  <info@fallabs.com>

	* ktserver.c (run): check to the number of databases was added.

	* ktulog.h (UpdateLogger::Reader::open): a bug related to ID detection was fixed.

	- Release: 0.9.52

2011-08-17  FAL Labs  <info@fallabs.com>

	* ktulog.h (UpdateLogger::Reader::open): a bug related to unsupported files was fixed.

	- Release: 0.9.51

2011-07-01  FAL Labs  <info@fallabs.com>

	* ktsocket.cc (waitsocket): the poll mode was added.

	- Release: 0.9.50

2011-06-28  FAL Labs  <info@fallabs.com>

	* ktplugservmemc.cc (Worker::do_queue_set): the delimiter was changed.

	- Release: 0.9.49

2011-06-22  FAL Labs  <info@fallabs.com>

	* ktdbext.h (MapReduce::FlushThread): new class.

	* ktdbext.h (MapReduce::cache_flush): parallel mode was added.

	- Release: 0.9.48

2011-06-17  FAL Labs  <info@fallabs.com>

	* kcutil.h (maskthreadsignal): new function.

	* ktthserv.h (TaskQueueImpl::do_start, TaskQueueImpl::do_finish): new functions.

	* ktplugservmmc.cc (Worker::configure): "tout" and "th" parameters were added.

	- Release: 0.9.47

2011-06-12  FAL Labs  <info@fallabs.com>

	* ktplugservmmc.cc (Worker::do_queue_get): signal handler was modified.

	* cmdcommon.h (DEFTHNUM): modified from 8 to 16.

	- Release: 0.9.46

2011-06-11  FAL Labs  <info@fallabs.com>

	* ktsocket.cc (_init_func): new function.

	* kttimeddb.cc (TimedDB::Curosr::seize, TimedDB::seize): new functions.

	* ktremotedb.cc (RemoteDB::Curosr::seize, RemoteDB::seize): new functions.

	* ktplugservmmc.cc (Worker::do_queue_set, Worker::do_queue_get): new functions.

	- Release: 0.9.45

2011-06-11  FAL Labs  <info@fallabs.com>

	* ktserver (Worker::process): "WAIT" and "SIGNAL" parameters were added.

	* ktserver (Worker::do_void): new function.

	* ktserver (proc): handling condition variable was added.

	* ktremotedb.h (RemoteDB::set_signal_waiting, RemoteDB::set_signal_sending): new functions.

	- Release: 0.9.44

2011-05-19  FAL Labs  <info@fallabs.com>

	* kttimeddb.h (TimedDB::scan_parallel): new function.

	* ktdbext.h (MapReduce::execute): parallel options were added.

	* ktdbext.h (MapReduce::tune_thread): new function.

	- Release: 0.9.43

2011-05-17  FAL Labs  <info@fallabs.com>

	* kttimeddb.h (TimedDB::increment): "orig" parameter was added.

	* ktserver.cc (Worker::do_increment): "orig" parameter was added.

	- Release: 0.9.42

2011-05-03  FAL Labs  <info@fallabs.com>

	* ktdbext.h (MapReduce::map): MapEmitter was integrated.

	* ktdbext.h (MapReduce::emit): new function.

	* ktdbext.h (MapReduce::execute): a bug of wrong handling of empty databases was fixed.

	- Release: 0.9.41

2011-04-22  FAL Labs  <info@fallabs.com>

	* kttimeddb.h (TimedDB::recover): expiration was edded.

	- Release: 0.9.40

2011-03-22  FAL Labs  <info@fallabs.com>

	* ktplugservmemc.cc (Worker::do_get): a bug retrieving a useless record was fixed.

	* myscript.cc (ScriptProcessor::call): a bug related to the nil character was fixed.

	* myscript.cc (db_get_bulk): a bug related to the nil character was fixed.

	- Release: 0.9.39

2011-03-06  FAL Labs  <info@fallabs.com>

	* ktplugdbvoid.cc (VoidDB::log): new function.

	- Release: 0.9.38

2011-02-25  FAL Labs  <info@fallabs.com>

	* kctimeddb.h (TimedDB::occupy): new function.

	* kttimeddb.h (TimedDB::dump_snapshot_atomic): fork mechanism was modified.

	* myscript.cc (db_occupy): new function.


kyototycoon/ktutilserv.cc
/*************************************************************************************************
 * The testing implementations using the server tool kit
 *                                                               Copyright (C) 2009-2012 FAL Labs
 * This file is part of Kyoto Tycoon.
 * This program is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version
 * 3 of the License, or any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this program.
 * If not, see <http://www.gnu.org/licenses/>.
 *************************************************************************************************/


#include "cmdcommon.h"


// global variables
const char* g_progname;                  // program name
kt::ServerSocket *g_servsock;            // running server socket
kt::Poller *g_poller;                    // running poller
kt::ThreadedServer* g_thserv;            // running threaded server
kt::HTTPServer* g_httpserv;              // running HTTP server
kt::RPCServer* g_rpcserv;                // running RPC server


// function prototypes
int main(int argc, char** argv);
static void usage();
static void killserver(int signum);
static int32_t runecho(int argc, char** argv);
static int32_t runmtecho(int argc, char** argv);
static int32_t runhttp(int argc, char** argv);
static int32_t runrpc(int argc, char** argv);
static int32_t procecho(const char* host, int32_t port, double tout);
static int32_t procmtecho(const char* host, int32_t port, double tout, int32_t thnum,
                          uint32_t logkinds);
static int32_t prochttp(const char* base,
                        const char* host, int32_t port, double tout, int32_t thnum,
                        uint32_t logkinds);
static int32_t procrpc(const char* host, int32_t port, double tout, int32_t thnum,
                       uint32_t logkinds);


// main routine
int main(int argc, char** argv) {
  g_progname = argv[0];
  kc::setstdiobin();
  kt::setkillsignalhandler(killserver);
  if (argc < 2) usage();
  int32_t rv = 0;
  if (!std::strcmp(argv[1], "echo")) {
    rv = runecho(argc, argv);
  } else if (!std::strcmp(argv[1], "mtecho")) {
    rv = runmtecho(argc, argv);
  } else if (!std::strcmp(argv[1], "http")) {
    rv = runhttp(argc, argv);
  } else if (!std::strcmp(argv[1], "rpc")) {
    rv = runrpc(argc, argv);
  } else if (!std::strcmp(argv[1], "version") || !std::strcmp(argv[1], "--version")) {
    printversion();
    rv = 0;
  } else {
    usage();
  }
  return rv;
}


// print the usage and exit
static void usage() {
  eprintf("%s: testing implementations using the server tool kit\n", g_progname);
  eprintf("\n");
  eprintf("usage:\n");
  eprintf("  %s echo [-host str] [-port num] [-tout num]\n", g_progname);
  eprintf("  %s mtecho [-host str] [-port num] [-tout num] [-th num] [-li|-ls|-le|-lz]\n",
          g_progname);
  eprintf("  %s http [-host str] [-port num] [-tout num] [-th num] [-li|-ls|-le|-lz]"
          " [basedir]\n", g_progname);
  eprintf("  %s rpc [-host str] [-port num] [-tout num] [-th num] [-li|-ls|-le|-lz]\n",
          g_progname);
  eprintf("\n");
  std::exit(1);
}


// kill the running server
static void killserver(int signum) {
  oprintf("%s: catched the signal %d\n", g_progname, signum);
  if (g_servsock) {
    g_servsock->abort();
    g_servsock = NULL;
  }
  if (g_poller) {
    g_poller->abort();
    g_poller = NULL;
  }
  if (g_thserv) {
    g_thserv->stop();
    g_thserv = NULL;
  }
  if (g_httpserv) {
    g_httpserv->stop();
    g_httpserv = NULL;
  }
  if (g_rpcserv) {
    g_rpcserv->stop();
    g_rpcserv = NULL;
  }
}


// parse arguments of echo command
static int32_t runecho(int argc, char** argv) {
  bool argbrk = false;
  const char* host = NULL;
  int32_t port = kt::DEFPORT;
  double tout = DEFTOUT;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-host")) {
        if (++i >= argc) usage();
        host = argv[i];
      } else if (!std::strcmp(argv[i], "-port")) {
        if (++i >= argc) usage();
        port = kc::atoi(argv[i]);
      } else if (!std::strcmp(argv[i], "-tout")) {
        if (++i >= argc) usage();
        tout = kc::atof(argv[i]);
      } else {
        usage();
      }
    } else {
      argbrk = true;
      usage();
    }
  }
  if (port < 1) usage();
  int32_t rv = procecho(host, port, tout);
  return rv;
}


// parse arguments of mtecho command
static int32_t runmtecho(int argc, char** argv) {
  bool argbrk = false;
  const char* host = NULL;
  int32_t port = kt::DEFPORT;
  double tout = DEFTOUT;
  int32_t thnum = DEFTHNUM;
  uint32_t logkinds = kc::UINT32MAX;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-host")) {
        if (++i >= argc) usage();
        host = argv[i];
      } else if (!std::strcmp(argv[i], "-port")) {
        if (++i >= argc) usage();
        port = kc::atoi(argv[i]);
      } else if (!std::strcmp(argv[i], "-tout")) {
        if (++i >= argc) usage();
        tout = kc::atof(argv[i]);
      } else if (!std::strcmp(argv[i], "-th")) {
        if (++i >= argc) usage();
        thnum = kc::atof(argv[i]);
      } else if (!std::strcmp(argv[i], "-li")) {
        logkinds = kt::ThreadedServer::Logger::INFO | kt::ThreadedServer::Logger::SYSTEM |
            kt::ThreadedServer::Logger::ERROR;
      } else if (!std::strcmp(argv[i], "-ls")) {
        logkinds = kt::ThreadedServer::Logger::SYSTEM | kt::ThreadedServer::Logger::ERROR;
      } else if (!std::strcmp(argv[i], "-le")) {
        logkinds = kt::ThreadedServer::Logger::ERROR;
      } else if (!std::strcmp(argv[i], "-lz")) {
        logkinds = 0;
      } else {
        usage();
      }
    } else {
      argbrk = true;
      usage();
    }
  }
  if (port < 1 || thnum < 1) usage();
  if (thnum > THREADMAX) thnum = THREADMAX;
  int32_t rv = procmtecho(host, port, tout, thnum, logkinds);
  return rv;
}


// parse arguments of http command
static int32_t runhttp(int argc, char** argv) {
  bool argbrk = false;
  const char* host = NULL;
  const char* base = NULL;
  int32_t port = kt::DEFPORT;
  double tout = DEFTOUT;
  int32_t thnum = DEFTHNUM;
  uint32_t logkinds = kc::UINT32MAX;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-host")) {
        if (++i >= argc) usage();
        host = argv[i];
      } else if (!std::strcmp(argv[i], "-port")) {
        if (++i >= argc) usage();
        port = kc::atoi(argv[i]);
      } else if (!std::strcmp(argv[i], "-tout")) {
        if (++i >= argc) usage();
        tout = kc::atof(argv[i]);
      } else if (!std::strcmp(argv[i], "-th")) {
        if (++i >= argc) usage();
        thnum = kc::atof(argv[i]);
      } else if (!std::strcmp(argv[i], "-li")) {
        logkinds = kt::HTTPServer::Logger::INFO | kt::HTTPServer::Logger::SYSTEM |
            kt::HTTPServer::Logger::ERROR;
      } else if (!std::strcmp(argv[i], "-ls")) {
        logkinds = kt::HTTPServer::Logger::SYSTEM | kt::HTTPServer::Logger::ERROR;
      } else if (!std::strcmp(argv[i], "-le")) {
        logkinds = kt::HTTPServer::Logger::ERROR;
      } else if (!std::strcmp(argv[i], "-lz")) {
        logkinds = 0;
      } else {
        usage();
      }
    } else if (!base) {
      argbrk = true;
      base = argv[i];
    } else {
      usage();
    }
  }
  if (port < 1 || thnum < 1) usage();
  if (thnum > THREADMAX) thnum = THREADMAX;
  int32_t rv = prochttp(base, host, port, tout, thnum, logkinds);
  return rv;
}


// parse arguments of rpc command
static int32_t runrpc(int argc, char** argv) {
  bool argbrk = false;
  const char* host = NULL;
  int32_t port = kt::DEFPORT;
  double tout = DEFTOUT;
  int32_t thnum = DEFTHNUM;
  uint32_t logkinds = kc::UINT32MAX;
  for (int32_t i = 2; i < argc; i++) {
    if (!argbrk && argv[i][0] == '-') {
      if (!std::strcmp(argv[i], "--")) {
        argbrk = true;
      } else if (!std::strcmp(argv[i], "-host")) {
        if (++i >= argc) usage();
        host = argv[i];
      } else if (!std::strcmp(argv[i], "-port")) {
        if (++i >= argc) usage();
        port = kc::atoi(argv[i]);
      } else if (!std::strcmp(argv[i], "-tout")) {
        if (++i >= argc) usage();
        tout = kc::atof(argv[i]);
      } else if (!std::strcmp(argv[i], "-th")) {
        if (++i >= argc) usage();
        thnum = kc::atof(argv[i]);
      } else if (!std::strcmp(argv[i], "-li")) {
        logkinds = kt::RPCServer::Logger::INFO | kt::RPCServer::Logger::SYSTEM |
            kt::RPCServer::Logger::ERROR;
      } else if (!std::strcmp(argv[i], "-ls")) {
        logkinds = kt::RPCServer::Logger::SYSTEM | kt::RPCServer::Logger::ERROR;
      } else if (!std::strcmp(argv[i], "-le")) {
        logkinds = kt::RPCServer::Logger::ERROR;
      } else if (!std::strcmp(argv[i], "-lz")) {
        logkinds = 0;
      } else {
        usage();
      }
    } else {
      argbrk = true;
      usage();
    }
  }
  if (port < 1 || thnum < 1) usage();
  if (thnum > THREADMAX) thnum = THREADMAX;
  int32_t rv = procrpc(host, port, tout, thnum, logkinds);
  return rv;
}


// perform echo command
static int32_t procecho(const char* host, int32_t port, double tout) {
  std::string addr = "";
  if (host) {
    addr = kt::Socket::get_host_address(host);
    if (addr.empty()) {
      eprintf("%s: %s: unknown host\n", g_progname, host);
      return 1;
    }
  }
  std::string expr = kc::strprintf("%s:%d", addr.c_str(), port);
  kt::ServerSocket serv;
  if (!serv.open(expr)) {
    eprintf("%s: server: open error: %s\n", g_progname, serv.error());
    return 1;
  }
  bool err = false;
  kt::Poller poll;
  if (!poll.open()) {
    eprintf("%s: poller: open error: %s\n", g_progname, poll.error());
    err = true;
  }
  g_servsock = &serv;
  g_poller = &poll;
  oprintf("%s: started: %s\n", g_progname, serv.expression().c_str());
  serv.set_event_flags(kt::Pollable::EVINPUT);
  if (!poll.deposit(&serv)) {
    eprintf("%s: poller: deposit error: %s\n", g_progname, poll.error());
    err = true;
  }
  while (g_servsock) {
    if (poll.wait()) {
      kt::Pollable* event;
      while ((event = poll.next()) != NULL) {
        if (event == &serv) {
          kt::Socket* sock = new kt::Socket;
          sock->set_timeout(tout);
          if (serv.accept(sock)) {
            oprintf("%s: connected: %s\n", g_progname, sock->expression().c_str());
            sock->set_event_flags(kt::Pollable::EVINPUT);
            if (!poll.deposit(sock)) {
              eprintf("%s: poller: deposit error: %s\n", g_progname, poll.error());
              err = true;
            }
          } else {
            eprintf("%s: server: accept error: %s\n", g_progname, serv.error());
            err = true;
            delete[] sock;
          }
          serv.set_event_flags(kt::Pollable::EVINPUT);
          if (!poll.undo(&serv)) {
            eprintf("%s: poller: undo error: %s\n", g_progname, poll.error());
            err = true;
          }
        } else {
          kt::Socket* sock = (kt::Socket*)event;
          char line[LINEBUFSIZ];
          if (sock->receive_line(line, sizeof(line))) {
            oprintf("%s: (%s): %s\n", g_progname, sock->expression().c_str(), line);
            if (!kc::stricmp(line, "/quit")) {
              if (!sock->printf("> Bye!\n")) {
                eprintf("%s: socket: printf error: %s\n", g_progname, sock->error());
                err = true;
              }
              oprintf("%s: closing: %s\n", g_progname, sock->expression().c_str());
              if (!poll.withdraw(sock)) {
                eprintf("%s: poller: withdraw error: %s\n", g_progname, poll.error());
                err = true;
              }
              if (!sock->close()) {
                eprintf("%s: socket: close error: %s\n", g_progname, poll.error());
                err = true;
              }
              delete sock;
            } else {
              if (!sock->printf("> %s\n", line)) {
                eprintf("%s: socket: printf error: %s\n", g_progname, sock->error());
                err = true;
              }
              serv.set_event_flags(kt::Pollable::EVINPUT);
              if (!poll.undo(sock)) {
                eprintf("%s: poller: undo error: %s\n", g_progname, poll.error());
                err = true;
              }
            }
          } else {
            oprintf("%s: closed: %s\n", g_progname, sock->expression().c_str());
            if (!poll.withdraw(sock)) {
              eprintf("%s: poller: withdraw error: %s\n", g_progname, poll.error());
              err = true;
            }
            if (!sock->close()) {
              eprintf("%s: socket: close error: %s\n", g_progname, poll.error());
              err = true;
            }
            delete sock;
          }
        }
      }
    } else {
      eprintf("%s: poller: wait error: %s\n", g_progname, poll.error());
      err = true;
    }
  }
  g_poller = NULL;
  if (poll.flush()) {
    kt::Pollable* event;
    while ((event = poll.next()) != NULL) {
      if (event != &serv) {
        kt::Socket* sock = (kt::Socket*)event;
        oprintf("%s: discarded: %s\n", g_progname, sock->expression().c_str());
        if (!poll.withdraw(sock)) {
          eprintf("%s: poller: withdraw error: %s\n", g_progname, poll.error());
          err = true;
        }
        if (!sock->close()) {
          eprintf("%s: socket: close error: %s\n", g_progname, poll.error());
          err = true;
        }
        delete sock;
      }
    }
  } else {
    eprintf("%s: poller: flush error: %s\n", g_progname, poll.error());
    err = true;
  }
  oprintf("%s: finished: %s\n", g_progname, serv.expression().c_str());
  if (!poll.close()) {
    eprintf("%s: poller: close error: %s\n", g_progname, poll.error());
    err = true;
  }
  if (!serv.close()) {
    eprintf("%s: server: close error: %s\n", g_progname, serv.error());
    err = true;
  }
  return err ? 1 : 0;
}


// perform mtecho command
static int32_t procmtecho(const char* host, int32_t port, double tout, int32_t thnum,
                          uint32_t logkinds) {
  std::string addr = "";
  if (host) {
    addr = kt::Socket::get_host_address(host);
    if (addr.empty()) {
      eprintf("%s: %s: unknown host\n", g_progname, host);
      return 1;
    }
  }
  std::string expr = kc::strprintf("%s:%d", addr.c_str(), port);
  kt::ThreadedServer serv;
  kt::ThreadedServer::Logger* logger = stdlogger(g_progname, &std::cout);
  class Worker : public kt::ThreadedServer::Worker {
   private:
    bool process(kt::ThreadedServer* serv, kt::ThreadedServer::Session* sess) {
      bool keep = false;
      char line[LINEBUFSIZ];
      if (sess->receive_line(line, sizeof(line))) {
        if (!kc::stricmp(line, "/quit")) {
          sess->printf("> Bye!\n");
        } else {
          class Data : public kt::ThreadedServer::Session::Data {
           public:
            time_t t;
          };
          Data* data = (Data*)sess->data();
          if (!data) {
            data = new Data;
            sess->set_data(data);
            data->t = kc::time();
          }
          sess->printf("> %s\n", line);
          serv->log(kt::ThreadedServer::Logger::INFO, "(%s): id=%llu thid=%u time=%lld msg=%s",
                    sess->expression().c_str(), (unsigned long long)sess->id(),
                    (unsigned)sess->thread_id(), (long long)(kc::time() - data->t), line);
          keep = true;
        }
      }
      return keep;
    }
  };
  Worker worker;
  bool err = false;
  serv.set_network(expr, tout);
  serv.set_logger(logger, logkinds);
  serv.set_worker(&worker, thnum);
  g_thserv = &serv;
  serv.log(kt::ThreadedServer::Logger::SYSTEM, "================ [START]");
  if (serv.start()) {
    if (serv.finish()) err = true;
  } else {
    err = true;
  }
  serv.log(kt::ThreadedServer::Logger::SYSTEM, "================ [FINISH]");
  return err ? 1 : 0;
}


// perform http command
static int32_t prochttp(const char* base,
                        const char* host, int32_t port, double tout, int32_t thnum,
                        uint32_t logkinds) {
  if (!base) base = kc::File::CDIRSTR;
  std::string baseabs = kc::File::absolute_path(base);
  if (baseabs.empty()) {
    eprintf("%s: %s: unknown directory\n", g_progname, base);
    return 1;
  }
  std::string addr = "";
  if (host) {
    addr = kt::Socket::get_host_address(host);
    if (addr.empty()) {
      eprintf("%s: %s: unknown host\n", g_progname, host);
      return 1;
    }
  }
  std::string expr = kc::strprintf("%s:%d", addr.c_str(), port);
  kt::HTTPServer serv;
  kt::HTTPServer::Logger* logger = stdlogger(g_progname, &std::cout);
  class Worker : public kt::HTTPServer::Worker {
   public:
    explicit Worker(const std::string& base) : base_(base) {}
   private:
    int32_t process(kt::HTTPServer* serv, kt::HTTPServer::Session* sess,
                    const std::string& path, kt::HTTPClient::Method method,
                    const std::map<std::string, std::string>& reqheads,
                    const std::string& reqbody,
                    std::map<std::string, std::string>& resheads,
                    std::string& resbody,
                    const std::map<std::string, std::string>& misc) {
      const char* url = kt::strmapget(misc, "url");
      int32_t code = -1;
      const std::string& lpath = kt::HTTPServer::localize_path(path);
      std::string apath = base_ + (lpath.empty() ? "" : kc::File::PATHSTR) + lpath;
      bool dir = kc::strbwm(path.c_str(), "/");
      if (method == kt::HTTPClient::MGET) {
        kc::File::Status sbuf;
        if (kc::File::status(apath, &sbuf)) {
          if (dir && sbuf.isdir) {
            const std::string& ipath = apath + kc::File::PATHSTR + "index.html";
            if (kc::File::status(ipath)) {
              apath = ipath;
              sbuf.isdir = false;
            }
          }
          if (sbuf.isdir) {
            if (dir) {
              std::vector<std::string> files;
              if (kc::File::read_directory(apath, &files)) {
                code = 200;
                resheads["content-type"] = "text/html";
                kc::strprintf(&resbody, "<html>\n");
                kc::strprintf(&resbody, "<body>\n");
                kc::strprintf(&resbody, "<ul>\n");
                kc::strprintf(&resbody, "<li><a href=\"./\">./</a></li>\n");
                kc::strprintf(&resbody, "<li><a href=\"../\">../</a></li>\n");
                kc::strprintf(&resbody, "</ul>\n");
                kc::strprintf(&resbody, "<ul>\n");
                std::sort(files.begin(), files.end());
                std::vector<std::string>::iterator it = files.begin();
                std::vector<std::string>::iterator itend = files.end();
                while (it != itend) {
                  if (*it != kc::File::CDIRSTR && *it != kc::File::PDIRSTR) {
                    std::string cpath = apath + kc::File::PATHSTR + *it;
                    if (kc::File::status(cpath, &sbuf)) {
                      char* ubuf = kc::urlencode(it->data(), it->size());
                      char* xstr = kt::xmlescape(it->c_str());
                      const char* dsuf = sbuf.isdir ? "/" : "";
                      kc::strprintf(&resbody, "<li><a href=\"%s%s\">%s%s</a></li>\n",
                                    ubuf, dsuf, xstr, dsuf);
                      delete[] xstr;
                      delete[] ubuf;
                    }
                  }
                  ++it;
                }
                kc::strprintf(&resbody, "</ul>\n");
                kc::strprintf(&resbody, "</body>\n");
                kc::strprintf(&resbody, "</html>\n");
              } else {
                code = 403;
                resheads["content-type"] = "text/plain";
                kc::strprintf(&resbody, "%s\n", kt::HTTPServer::status_name(code));
              }
            } else {
              code = 301;
              resheads["content-type"] = "text/plain";
              resheads["location"] = (url ? url : path) + '/';
              kc::strprintf(&resbody, "%s\n", kt::HTTPServer::status_name(code));
            }
          } else {
            int64_t size;
            char* buf = kc::File::read_file(apath, &size, 256LL << 20);
            if (buf) {
              code = 200;
              const char* type = kt::HTTPServer::media_type(path);
              if (type) resheads["content-type"] = type;
              resbody.append(buf, size);
              delete[] buf;
            } else {
              code = 403;
              resheads["content-type"] = "text/plain";
              kc::strprintf(&resbody, "%s\n", kt::HTTPServer::status_name(code));
            }
          }
        } else {
          code = 404;
          resheads["content-type"] = "text/plain";
          kc::strprintf(&resbody, "%s\n", kt::HTTPServer::status_name(code));
        }
      } else {
        code = 403;
        resheads["content-type"] = "text/plain";
        kc::strprintf(&resbody, "%s\n", kt::HTTPServer::status_name(code));
      }
      return code;
    }
    std::string base_;
  };
  Worker worker(baseabs);
  bool err = false;
  serv.set_network(expr, tout);
  serv.set_logger(logger, logkinds);
  serv.set_worker(&worker, thnum);
  g_httpserv = &serv;
  serv.log(kt::HTTPServer::Logger::SYSTEM, "================ [START]");
  if (serv.start()) {
    if (serv.finish()) err = true;
  } else {
    err = true;
  }
  serv.log(kt::HTTPServer::Logger::SYSTEM, "================ [FINISH]");
  return err ? 1 : 0;
}


// perform rpc command
static int32_t procrpc(const char* host, int32_t port, double tout, int32_t thnum,
                       uint32_t logkinds) {
  std::string addr = "";
  if (host) {
    addr = kt::Socket::get_host_address(host);
    if (addr.empty()) {
      eprintf("%s: %s: unknown host\n", g_progname, host);
      return 1;
    }
  }
  std::string expr = kc::strprintf("%s:%d", addr.c_str(), port);
  kt::RPCServer serv;
  kt::RPCServer::Logger* logger = stdlogger(g_progname, &std::cout);
  class Worker : public kt::RPCServer::Worker {
   private:
    kt::RPCClient::ReturnValue process(kt::RPCServer* serv, kt::RPCServer::Session* sess,
                                       const std::string& name,
                                       const std::map<std::string, std::string>& inmap,
                                       std::map<std::string, std::string>& outmap) {
      outmap.insert(inmap.begin(), inmap.end());
      return kt::RPCClient::RVSUCCESS;
    }
  };
  Worker worker;
  bool err = false;
  serv.set_network(expr, tout);
  serv.set_logger(logger, logkinds);
  serv.set_worker(&worker, thnum);
  g_rpcserv = &serv;
  serv.log(kt::RPCServer::Logger::SYSTEM, "================ [START]");
  if (serv.start()) {
    if (serv.finish()) err = true;
  } else {
    err = true;
  }
  serv.log(kt::RPCServer::Logger::SYSTEM, "================ [FINISH]");
  return err ? 1 : 0;
}



// END OF FILE
